### ISSUE TITLE
set_epoch() is only called on the train dataloader. impossible to shuffle val/test dataset in DDP.

### ISSUE THREAD

**Issue by zplizzi** [PyTorchLightning/pytorch-lightning#10342]
On [this line](https://github.com/PyTorchLightning/pytorch-lightning/blob/a967b6eba0556943ad1d7c2a8dc41f1da4f68b2d/pytorch_lightning/loops/fit_loop.py#L216), `set_epoch()` is called on the train dataloader. This allows shuffling to work properly when using DDP. See #224 for context on why this was necessary.

However, `set_epoch()` is not called for the test or val dataloaders. Typically, these aren't shuffled, so it's fine. But sometimes the user wants to shuffle these datasets also, and as is, that doesn't work (and there's no indication to the user why it's not working). I can't see any reason not to call `set_epoch()` on these other dataloaders also.

**Comment by tchaton** [comment]
Dear @zplizzi,

Would you be willing to contribute a fix depending on the sampler / shuffle value ?

Best,
T.C

**Comment by bibhabasumohapatra** [comment]
"_impossible to shuffle val/test dataset in DDP_"
I wanted to say that ,  we should not be able to shuffle val/test ,  even when it is required in practice?

**Comment by bsridatta** [comment]
Is anyone already working on this? I would like to contribute.

**Comment by zplizzi** [comment]
Go for it!

**Comment by bsridatta** [comment]
Sure, I shall work on this today, thanks!

**Comment by ananthsub** [comment]
@zplizzi out of curiosity could you describe why ordering matters for your validation/test sets?

**Comment by bibhabasumohapatra** [comment]
> "_impossible to shuffle val/test dataset in DDP_"
> I wanted to say that , is it required to be able to shuffle val/test , even when it is required in practice?

@ananthsub    i had already asked ? same doubt is it required?

**Comment by zplizzi** [comment]
In this case, I am visualizing the first batch of validation samples, and I would like my visualization to not always be of the same batch of frames. But more generally, it seems wrong to allow the user to specify `shuffle=True` and then silently not shuffle. Why not just let the user get the behavior they ask for (when the effort required is trivial, as it is here)?

**Comment by xuanyang19** [comment]
How could I call set_epoch() on DistributedSampler in pytorch lightning? Although I set distributedsampler(shuffle = True), the train index is always the same at every epoch. What should I do to make the train index become different at every epoch?
my dataloader:
DataLoader(self.trainset,sampler=BatchSampler(DistributedSampler(self.trainset, shuffle=True), batch_size=self.batch_size,drop_last=False), num_workers=self.num_workers)
